common/message_batcher/src/lib.rs

```diff
@@ -0,0 +1,439 @@
+    /// Returns true, if the message belongs to the current batch.
+    fn belongs_to_batch(&self, message: &T, message_batch: &MessageBatch<T>) -> bool;
+}
+
+/// Inputs to the `MessageBatcher`.
+pub enum Input<T> {
+    /// A message was received.
+    Message {
+        /// Time the message has been received.
+        received_at: Timestamp,
+
+        /// The message itself.
+        message: T,
+    },
+
+    /// Notify the `MessageBatcher` about an expired timer, requested through `Output::NotifyAt`.
+    ///
+    /// Allows the `MessageBatcher` to close a batch when the current time window has expired.
+    Notify {
+        /// The current system time.
+        now: Timestamp,
+    },
+
+    /// Will flush the current batch. This is used upon termination.
+    Flush,
+}
+
+/// Outputs from the `MessageBatcher`. These inform the imperative shell to perform some actions on behalf
+/// of the `MessageBatcher`.
+#[derive(Debug, PartialEq)]
+pub enum Output<T> {
+    /// Informs the imperative shell to send an `Input::Notify` at (or slightly after) the specified timestamp.
+    NotifyAt(Timestamp),
+
+    /// Informs the imperative shell to send the message batch out.
+    MessageBatch(MessageBatch<T>),
+}
+
+/// The MessageBatcher's internal state / configuration.
+pub struct MessageBatcher<T> {
+    /// Maximum number of messages per batch.
+    max_batch_size: usize,
+
+    /// The maximum age of a batch.
+    ///
+    /// Age of a batch is the elapsed time since `current_batch.opened_at`.
+    max_batch_age: chrono::Duration,
+
+    /// The decisions whether a message belongs to the current batch or not.
+    batching_criteria: Vec<Box<dyn BatchingCriterion<T>>>,
+
+    current_batch: Option<MessageBatch<T>>,
+}
+
+impl<T> MessageBatcher<T> {
+    pub fn new(max_batch_size: usize, max_batch_age: chrono::Duration) -> Self {
+        assert!(max_batch_size > 0);
+        Self {
+            max_batch_size,
+            max_batch_age,
+            batching_criteria: Vec::new(),
+            current_batch: None,
+        }
+    }
+
+    pub fn with_batching_criterion(
+        mut self,
+        batching_criterion: impl BatchingCriterion<T> + 'static,
+    ) -> Self {
+        self.batching_criteria.push(Box::new(batching_criterion));
+        self
+    }
+
+    pub fn handle(&mut self, input: Input<T>, outputs: &mut Vec<Output<T>>) {
+        match input {
+            Input::Message {
+                message,
+                received_at,
+            } => self.handle_message(message, received_at, outputs),
+            Input::Notify { now } => self.handle_notify(now, outputs),
+            Input::Flush => self.handle_flush(outputs),
+        }
+
+        // Inform imperative shell about when to send a `Input::Notify` message.
+        if let Some(batch_opened_at) = self.current_batch.as_ref().map(|batch| batch.opened_at) {
+            outputs.push(Output::NotifyAt(batch_opened_at + self.max_batch_age));
+        }
```

Comment:
This needs to be clarified.

So this is to schedule a timeout notification, on which the current batch will be closed.

But:
* The timing criteria is not clear. The notification event is computed after the *event* time. While the outer loop use the *system* time. At least this is my understanding.
* I would add a delta to acknowledge the fact that the messages from the source arrive with some delay: at `system-time = t + delta` then one can assume that all the messages sent at `event-time = t` are arrived.
* This code corresponds to what we call `sleep` outside the `select loop` i.e a criteria between the first and last message. What about a `sleep` inside the `select loop` i.e batching messages that have been emitted one right after the other?
* Why emit this output here rather than in `self.handle_message()` when a new batch is started? Here, we will trigger a timeout for every message.